qbraid/programs/gate_model/qasm2.py

```diff
@@ -38,58 +35,50 @@ def __init__(self, program: Qasm2StringType):
         if not isinstance(program, Qasm2String):
             raise ProgramTypeError(message=f"Expected 'str' object, got '{type(program)}'.")
 
-    def parsed(self) -> Program:
-        """Parse the program string."""
-        return parse(self._program)
-
-    def _get_bits(self, bit_type: str) -> list[tuple[str, int]]:
-        """Return the number of qubits or classical bits in the circuit.
-
-        Args:
-            bit_type: either "q" or "c" for qubits or classical bits, respectively.
+        self._module = pyqasm.loads(program)
+        self._module.validate()
 
-        Returns:
-            list[tuple[str, int]]: A list of qubits or classical bits. Each bit is
-                a tuple of (register name, index)
-        """
-        matches = re.findall(rf"{bit_type}reg (\w+)\[(\d+)\];", self.program)
+    @property
+    def qubits(self) -> dict[str, int]:
+        """Return the qubits acted upon by the operations in this circuit"""
+        return self._module._qubit_registers
 
-        result = []
-        for match in matches:
-            var_name = match[0]
-            n = int(match[1])
-            result.extend([(var_name, i) for i in range(n)])
-        return result
+    @property
+    def module(self) -> pyqasm.Module:
+        """Return the pyqasm module."""
+        return self._module
 
     @property
-    def qubits(self) -> list[tuple[str, int]]:
-        """Use regex to extract all qreg definitions from the string"""
-        return self._get_bits("q")
+    def num_qubits(self) -> int:
+        """Return the number of qubits in the circuit."""
+        return self._module.num_qubits
 
     @property
     def num_clbits(self) -> int:
         """Return the number of classical bits in the circuit."""
-        return len(self._get_bits("c"))
+        return self._module.num_clbits
 
     @property
     def depth(self) -> int:
         """Return the circuit depth (i.e., length of critical path)."""
-        counts = {q: 0 for q in self.qubits}
-        program = self.parsed()
-        counts = depth(program.statements, counts)
-        return max(counts.values())
+        return self._module.depth()
 
     def _unitary(self) -> np.ndarray:
         """Return the unitary of the QASM"""
-        return transpiler.transpile(self.program, "cirq").unitary()
+        module = self._module.copy()
+        module.unroll()
+        return transpiler.transpile(str(module), "cirq").unitary()
```

Comment:
I know this is unrelated, and carries over from before, but we should probably just `raise NotImplemented` here because it is not guaranteed that the user will have `cirq` installed. This might impact some tests though so may have to account for that.